### PR TITLE
Extends bufferlist interface to calculate crc at arbitrary range and…

### PR DESCRIPTION
**Pull request by ifed01**
… refactors Checksummer to use that

Signed-off-by: Igor Fedotov ifedotov@mirantis.com

### CODE REVIEW

File: src/common/buffer.cc [hunk 1]
```diff
@@ -2278,6 +2278,26 @@ int buffer::list::write_fd_zero_copy(int fd) const
   return 0;
 }
 
+__u32 buffer::list::crc32c(unsigned off, size_t len, __u32 crc) const
+{
+  if (off == 0 && len == length()) {
+    return crc32c(crc); //fall back to full implementation that supports caching.
+  }
+  buffer::list::const_iterator it = begin();
+  it.advance(MIN(off, length()));
```
Comment (liewegas) [inline src/common/buffer.cc]: This advance() call is not free--it has to traverse the bufferlist's list of ptr's.  I was imagining a crc32c method that works directly on the iterator and uses get_ptr_and_advance().

**Comment by ifed01** [comment]
@liewegas Fixed. Made static method for bufferlist. Another option would bes to add crc32c func to iterator class but I'd prefer to have all crc32c implementation close to each other (i.e. within bufferlist class)

**Comment by liewegas** [comment]
i was originally thinking we could add crc32c to iterate and switch 
the list method to just call it.  not sure how that interacts with teh 
crc caching stuff, though.  (also not sure if we really want ot keep 
that or not...)

**Comment by ifed01** [comment]
@liewegas, not sure I understand what variant would you prefer. Where should basic crc32c method be implemented? As an iterator member? or bufferlist?

**Comment by liewegas** [comment]
like this: http://github.com/liewegas/ceph/commit/wip-buffer-crc32c 

should be easy for checksummer to consume.  probably deserves a unit test (for spanning bufferptrs, ensuring iterator advances the proper amount?)

on second thought not sure it makes sense for the list version to consume this because of the caching code.  but we should revisit whether the crc cache really makes sense...

**Comment by ifed01** [comment]
Closing this PR, new one has been opened:
https://github.com/ceph/ceph/pull/9449

**Pull request closed without merging**